src/Stack/Types/Config.hs

```diff
@@ -441,6 +443,21 @@ instance FromJSON ApplyGhcOptions where
             "everything" -> pure AGOEverything
             _ -> fail $ "Invalid ApplyGhcOptions: " ++ show t
 
+data AllowNewer
+  = AllowNewerAll
+  | AllowNewerNone
+  | AllowNewerPackages [String]
+  deriving (Show, Read, Eq, Ord)
+
+instance FromJSON AllowNewer where
+  parseJSON v =
+    case v of
+      Bool True -> pure AllowNewerAll
+      Bool False -> pure AllowNewerNone
+      String "all" -> pure AllowNewerAll
+      String "none" -> pure AllowNewerNone
+      _ -> AllowNewerPackages <$> parseJSON v
```

Comment:
I am wondering if, rather than the current catch-all `_`, `parseJSON v` should distinguish valid arrays from other `v` and then treat the residual invalid `v` differently.